source/OrderedTypes/ZornsLemma.lagda

```diff
@@ -0,0 +1,375 @@
+
+    ρ : Propositional-Resizing
+    ρ = EM-gives-PR lem
+
+    C : X → 𝓤 ̇
+    C x = resize ρ (B x) (B-is-prop x)
+
+    τ : Nat C B
+    τ x = from-resize ρ (B x) (B-is-prop x)
+
+    τ-is-equiv : (x : X) → is-equiv (τ x)
+    τ-is-equiv x = from-resize-is-equiv ρ (B x) (B-is-prop x)
+
+    X'' : 𝓤 ̇
+    X'' = Σ x ꞉ X , C x
+
+    e = X''       ≃⟨ NatΣ τ , NatΣ-is-equiv C B τ τ-is-equiv ⟩
+        X'        ≃⟨ ≃-sym (f' , f'-is-equiv) ⟩
+        Ordinal 𝓤 ■
+
+    the-type-of-ordinals-is-small : is-small (Ordinal 𝓤)
+    the-type-of-ordinals-is-small = X'' , e
+
+    absurd : 𝟘
+    absurd = the-type-of-ordinals-is-large the-type-of-ordinals-is-small
+
+has-maximal-element : (𝓤 ⊔ 𝓣) ̇
+has-maximal-element = ∃ x ꞉ X , ((y : X) → x ≪ y → x ＝ y)
+
+
+choice-function-gives-zorns-lemma
+ : Excluded-Middle
+ → poset-axioms (_≪_)
+ → all-chains-have-upper-bound
+ → (Σ ε ꞉ (𝓟 {𝓣 ⊔ 𝓤} X → X) , ((A : 𝓟 X) → is-inhabited A → ε A ∈ A))
+ → has-maximal-element
+
+choice-function-gives-zorns-lemma lem axioms chains cf = ¬¬Σ→∃ ¬¬max
+ where
+  dn : {𝓥 : Universe } → DNE 𝓥
+  dn {𝓥} = EM-gives-DNE lem
+  ¬¬Σ→∃ : {𝓤 : Universe} {A : X → 𝓤 ̇} → ¬ ¬ (Σ x ꞉ X , A x ) → (∃ x ꞉ X , A x)
+  ¬¬Σ→∃ {𝓤} {A} inside = dn _ ∥∥-is-prop (¬¬-functor ∣_∣ inside)
+  ¬¬max : ¬ ¬ has-maximal-element-strong
+  ¬¬max = (choice-function-gives-not-not-zorns-lemma lem axioms chains cf)
+
+axiom-of-choice-implies-zorns-lemma
+ : Axiom-of-Choice
+ → poset-axioms (_≪_)
+ → (all-chains-have-upper-bound → has-maximal-element)
+
+axiom-of-choice-implies-zorns-lemma ac (X-is-set , axioms-rest) = III
+ where
+  em : Excluded-Middle
+  em = Choice-gives-Excluded-Middle pe' ac
+
+  lifted-cf
+   : ∥ Lift (𝓤 ⊔ 𝓣) X ∥
+   → ∃ ε ꞉ (𝓟 (Lift (𝓤 ⊔ 𝓣) X) → (Lift (𝓤 ⊔ 𝓣) X))
+         , ((A : 𝓟 (Lift (𝓤 ⊔ 𝓣) X)) → is-inhabited A → ε A ∈ A)
+  lifted-cf =
+   Choice-gives-Choice₄ ac (Lift (𝓤 ⊔ 𝓣) X) (Lift-is-set (𝓤 ⊔ 𝓣) X X-is-set)
+
+  lower-cf
+   : Σ ε ꞉ (𝓟 (Lift (𝓤 ⊔ 𝓣) X) → (Lift (𝓤 ⊔ 𝓣) X))
+         , ((A : 𝓟 (Lift (𝓤 ⊔ 𝓣) X)) → is-inhabited A → ε A ∈ A)
+   →  Σ ε ꞉ (𝓟 X → X) , ((A : 𝓟 X) → is-inhabited A → ε A ∈ A)
+  lower-cf (ϵ , f) = (ϵ' , f')
+   where
+    ϵ' : 𝓟 X → X
+    ϵ' S = lower (ϵ (S ∘ lower))
+    inhab-trans : {A' : 𝓟 X} → is-inhabited A' → is-inhabited (A' ∘ lower)
+    inhab-trans {A'} isA' =
+     isA' >>= λ isA'' →
+      ∣ lift (𝓤 ⊔ 𝓣) (pr₁ isA'')
+       , transport (λ q → (A' q) holds) (ε-Lift (𝓤 ⊔ 𝓣) (pr₁ isA'')) (pr₂ isA'')∣
+    f' : (A' : 𝓟 X) → is-inhabited A' → ϵ' A' ∈ A'
+    f' A' A'-inhab = (f (A' ∘ lower) (inhab-trans {A'} A'-inhab))
+
+  choice-function : ∥ X ∥ → ∃ ε ꞉ (𝓟 X → X) , ((A : 𝓟 X) → is-inhabited A → ε A ∈ A)
+  choice-function isX =  ∥∥-functor lower-cf (lifted-cf (∥∥-functor (lift (𝓤 ⊔ 𝓣)) isX))
+
+  I'
+   : all-chains-have-upper-bound
+   → ∃ ε ꞉ (𝓟 X → X) , ((A : 𝓟 X) → is-inhabited A → ε A ∈ A)
+   → has-maximal-element
+  I' chains = ∥∥-rec (∃-is-prop)
+          (choice-function-gives-zorns-lemma em (X-is-set , axioms-rest) chains)
```

Comment:
Indentation could be better here.

Reply:
I agree but I'm not sure how I would change it. Any further thoughts?

Reply:
I usually do this like this:
```
∥∥-rec 
 (∃-is-prop)
 (choice-function-gives-zorns-lemma em (X-is-set , axioms-rest) chains)
``` 
In general,
```
some-function
 argument1
 argument2
 ...
```
In more complicated situations,
```
some-function
 argument1
 (some-other-function
    argumenta
    argumentb)
 argument3
```

Reply:
You can fix this in a future pull request.